Add AdvertiseComponent and add it to vending machines

Visne opened the pull request:
<!-- The text between the arrows are comments - they will not be visible on your PR. -->
<!-- To automatically tag this PR, add the uppercase label(s) surrounded by brackets below, for example: [LABEL] -->

## About the PR <!-- Describe the Pull Request here. What does it change? What other things could this impact? -->
Adds AdvertiseComponent, a component to broadcast a message to chat at random intervals. This PR also adds it to vending machines so they can advertise their product.

![image](https://user-images.githubusercontent.com/39844191/113083312-49af7400-91dc-11eb-8e71-7a9b963986b8.png)

**AdvertiseComponent**
It's a component that sends a random message from the `pack` at a random interval, between `minWait` and `maxWait`. `pack` is a `AdvertisingPackPrototype`, which contains a list of advertisements. Advertising can be paused and resumed by using `AdvertiseComponent.Pause()` and `AdvertiseComponent.Resume()`.

**Vending Machines**
The component has been added to vending machines, and advertisements were taken from TG. When a vending machine is broken or unpowered, it stops advertising. There are a few vending machines without ads. I decided to print a warning when a vending machine does not have ads, which means there will be three warnings printed on saltern currently. I can remove this if it's annoying, but it seems nice to have ads for every vending machine. Not all of TGs ads are as funny, so tell me which ones I should remove.

**Changelog**
<!--
Here you can fill out a changelog that will automatically be added to the game when your PR is merged
There are 4 icons for changelog entries: add, remove, tweak, fix. I trust you can figure out the rest.

You can put your name after the :cl: symbol to change the name that shows in the changelog (otherwise it takes your GitHub username)
Like so: :cl: PJB

Generally, only put things in changelogs that players actually care about. Stuff like "Refactored X system, no changes should be visible" shouldn't be on a changelog.

For writing actual entries, don't consider the entry type suffix (e.g. add) to be "part" of the sentence:
bad: - add: a new tool for engineers
good: - add: added a new tool for engineers
-->

:cl:
- add: Added advertising to vending machines

Fixes #3273

PJB3005 commented on Content.Server/GameObjects/Components/AdvertiseComponent.cs:
```diff
@@ -0,0 +1,127 @@
+                return;
+            }
+
+            _advertisements = packPrototype.Advertisements.ToList();
```
Why copy this list? Can't you just keep a reference to the prototype?

PJB3005 commented on Content.Server/GameObjects/Components/AdvertiseComponent.cs:
```diff
@@ -0,0 +1,127 @@
+            if (string.IsNullOrEmpty(PackPrototypeId) || packPrototype == null)
+            {
+                // If there is no pack, log a warning and remove the component
+                Logger.Log(LogLevel.Warning, Owner.Name + " has " + Name + "Component but no advertisments pack.");
```
```suggestion
                Logger.Warning($"{Owner} has {Name} Component but no advertisments pack.");
```

Logging `Owner.Name` is pretty crappy because it doesn't have good debugging info like prototype ID. Just logging the entity's `ToString()` (what my suggestion does) DOES print prototype ID so makes debugging easier.

PJB3005 commented on Content.Server/GameObjects/Components/AdvertiseComponent.cs:
```diff
@@ -0,0 +1,127 @@
+            if (_advertisements.Count == 0)
+            {
+                // If no advertisements could be loaded, log a warning and remove component
+                Logger.Log(LogLevel.Warning, Owner.Name + " tried to load advertisements pack without ads.");
```
```suggestion
                Logger.Warning($"{Owner} tried to load advertisements pack without ads.");
```

Visne commented on Content.Server/GameObjects/Components/AdvertiseComponent.cs:
```diff
@@ -0,0 +1,127 @@
+        /// Minimum time to wait before saying a new ad, in ms.
+        /// </summary>
+        [field: DataField("minWait")]
+        private int MinWait { get; } = 40000;
```
Sure, do I keep them ints or make them floats to allow finer control? I assume ints would be enough.

SweptWasTaken commented on Resources/Prototypes/Catalog/VendingMachines/Advertisements/sec.yml:
```diff
@@ -0,0 +1,12 @@
+﻿- type: advertisementsPack
+  id: SecTech ads
+  advertisements:
+  - "Crack capitalist skulls!"
```
Remove this one I think

Visne replied:
Changed to "communist"

Visne commented:
> 
> 
> You should localize the advertisements instead.

Localization added and problems fixed. Ready for review again.

metalgearsloth commented on Content.Server/GameObjects/Components/VendingMachines/VendingMachineComponent.cs:
```diff
@@ -129,6 +129,18 @@ private void UpdatePower(PowerChangedMessage args)
             var state = args.Powered ? VendingMachineVisualState.Normal : VendingMachineVisualState.Off;
             TrySetVisualState(state);
+
+            // Pause/resume advertising if advertising component exists and not broken
+            if (!Owner.TryGetComponent(out AdvertiseComponent? advertiseComponent) || _broken) return;
+
+            if (Powered)
+            {
+                advertiseComponent.Resume();
+            }
+            else
+            {
+                advertiseComponent.Pause();
+            }
```
This is kinda dirty and should be an event or the likes to try and separate out the functionality.

Visne replied:
Will be kept like this as decided on Discord (https://discord.com/channels/310555209753690112/770682801607278632/832591047888601099).

metalgearsloth commented on Content.Server/GameObjects/Components/VendingMachines/VendingMachineComponent.cs:
```diff
@@ -244,6 +256,11 @@ public void OnBreak(BreakageEventArgs eventArgs)
             _broken = true;
             TrySetVisualState(VendingMachineVisualState.Broken);
+
+            if (Owner.TryGetComponent(out AdvertiseComponent? advertiseComponent))
+            {
+                advertiseComponent.Pause();
+            }
```
Same here.

metalgearsloth commented on Content.Shared/VendingMachines/VendingMachineInventoryPrototype.cs:
```diff
@@ -8,7 +8,7 @@
 namespace Content.Shared.VendingMachines
 {
-    [Serializable, NetSerializable, Prototype("vendingMachineInventory")]
+    [Serializable, NetSerializable, Prototype("VendingMachineInventory")]
```
I thought style guide was camelCase?

Visne replied:
No, https://hackmd.io/Awj5TvdXTaOmhU2XZOsdQg#Naming.

metalgearsloth replied:
See 
![image](https://user-images.githubusercontent.com/31366439/114891590-5a542100-9e4f-11eb-97cd-f5aac92cdaf4.png)

Visne replied:
You're correct, fixed.

metalgearsloth commented on Content.Server/GameObjects/Components/AdvertiseComponent.cs:
```diff
@@ -0,0 +1,140 @@
+    [RegisterComponent]
+    public class AdvertiseComponent : Component
+    {
+        [Dependency] private readonly IChatManager _chatManager = default!;
+        [Dependency] private readonly IRobustRandom _random = default!;
+        [Dependency] private readonly IPrototypeManager _prototypeManager = default!;
```
IChatManager + IRobustRandom can probably just be resolved directly given it only fires off every 40-80 seconds by default
IPrototypeManager is used once so just resolve it locally so you don't keep it around in memory

Normally we only store the member ones if they're being used frequently (and even then if you use an ECS approach you can just store them on a system instead).

Visne replied:
Fixed

metalgearsloth commented on Content.Server/GameObjects/Components/AdvertiseComponent.cs:
```diff
@@ -0,0 +1,140 @@
+        /// <summary>
+        /// Pause the advertising until <see cref="Resume"/> is called.
+        /// </summary>
+        public void Pause()
+        {
+            // Cancel current timer
+            _cancellationSource.Cancel();
+        }
```
Technically it's more of a "Stop" isn't it?

Visne replied:
For the person using the component it's ultimatelly the same as pausing and resuming, and I feel like it makes more sense this way. Can change it if you want though.

Visne commented:
This is ready for review again. Major changes since last review:

- Increased time delay until new ad is broadcasted
- Remove ads from department vending machines (only commercial vending machines will have ads)
- Changed initialization/resume to start timer without a minimum wait bound, so ads are more randomized when considering multiple vending machines.

metalgearsloth commented on Content.Server/GameObjects/Components/AdvertiseComponent.cs:
```diff
@@ -0,0 +1,144 @@
+                throw new PrototypeLoadException($"{Owner} should have minWait greater than or equal to maxWait for {Name}Component.");
+            }
+
+            // Start timer at initialization, without MinWait bound
+            RefreshTimer(false);
+        }
+
+        public override void OnRemove()
+        {
+            _cancellationSource.Cancel();
+            _cancellationSource.Dispose();
+
+            base.OnRemove();
+        }
```
Sorry I remembered one more thing that I made into a separate issue as I keep forgetting about it.

https://github.com/space-wizards/space-station-14/issues/3869

You should probably do these in Startup and Shutdown instead so if an admin can disable / enable advertising it's easy to do so.

metalgearsloth replied:
NVM turns out it's being deprecated

Qustinnus commented on Resources/Locale/en-US/advertisements/vending/sovietsoda.ftl:
```diff
@@ -0,0 +1,5 @@
+﻿advertisement-sovietsoda-1 = For Tsar and Country.
```
i feel like soviet soda wouldn't like the Tsar

Visne replied:
Yeah I honestly don't know who came up with that, can you come up with something better perhaps?

RemieRichards replied:
For Comrade and Country?
For Comrade and Party?

Visne replied:
Changed to "For comrade and country."
